### PR TITLE
Feat/implement service spec

### LINKED ISSUE #115: Add serviceTemplate
as part of https://github.com/aenix-io/etcd-operator/issues/109

```yaml
  serviceTemplate:
    metadata:
      labels:
        env: prod
      annotations:
        example.com/annotation: "true"
    spec: # core.v1.ServiceSpec Ready k8s type
```

### LINKED ISSUE #166: Rename services (breaking)
Rename services:
- `<clusterName>` --> `<clusterName>-headless`
- `<clusterName>-client` --> `<clusterName>`

Configure pods to access each other using `pod.service` hostname, not `pod.service.namespace.svc`

This change will require updating member names and certificates for existing installations

### PR DESCRIPTION
**Pull request by sergeyshevch**
Fixes #115 
Fixes #166

### CODE REVIEW

File: internal/controller/factory/metadata.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/aenix-io/etcd-operator/api/v1alpha1"
+)
+
+// mergeObjectMeta merges the metadata of an existing object with the metadata of a new object.
+// It copies the labels and annotations from the new object to the existing object.
+func mergeObjectMeta(old metav1.ObjectMeta, new v1alpha1.EmbeddedObjectMetadata) metav1.ObjectMeta {
```
Comment (Kirill-Garbar) [inline internal/controller/factory/metadata.go:11]: I would not name it merge if it just copies :)

**Review by Kirill-Garbar — APPROVED** [review]

File: internal/controller/factory/metadata.go (hunk 1 above)
Comment (sergeyshevch) [reply internal/controller/factory/metadata.go:11]: It's some sort of merge :) It merges objectMeta with specified embeddedObjectMetadata.

This function can be extended later with some more advanced logic (Such as propagation of EtcdCluster labels into all resources that we discussed earlier) So I suggest to leave this name

File: api/v1alpha1/etcdcluster_types.go [hunk 2]
```diff
@@ -242,6 +248,21 @@ type PodDisruptionBudgetSpec struct {
 	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
 }
 
+type ServiceSpec struct {
```
[the remark this answers is not in the document]
Comment (sircthulhu) [reply api/v1alpha1/etcdcluster_types.go:251]: btw there must be a validation hook that does make `metadata.name` immutable for both services

**Review by sircthulhu — CHANGES_REQUESTED** [review]

File: internal/controller/factory/metadata.go (hunk 1 above)
Comment (sergeyshevch) [reply internal/controller/factory/metadata.go:11]: It doesn't alter incoming parameters. It's a clean function. We deepcopy old medatata.

File: api/v1alpha1/etcdcluster_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergeyshevch) [reply api/v1alpha1/etcdcluster_types.go:251]: Then we also need to rename PodDisruptionBudget spec following this logic

**Review by hiddenmarten — APPROVED** [review]

**Review by sircthulhu — APPROVED** [review]
lgtm

**Review by Kirill-Garbar — APPROVED** [review]

**Review by kvaps — CHANGES_REQUESTED** [review]
Thanks! Also could you please update `dnsNames` in 

`/examples/manifests/etcdcluster-kamaji.yaml`
`/examples/manifests/etcdcluster-with-external-certificates.yaml`

**Review by hiddenmarten — APPROVED** [review]

**Review by kvaps — APPROVED** [review]

**Review by Kirill-Garbar — APPROVED** [review]

**Pull request merged**